src/util/haptics.ts

```diff
@@ -0,0 +1,12 @@
+/**
+ * Utility function that enables haptic feedback
+ */
+
+export function forwardHaptic(hapticType: string) {
```

Comment:
This is a no go, you're making all elements all of a sudden aware of external auth, hence your need of doing all the type changes. We should implement `forwardHaptic` as events. So it becomes `forwardHaptic(el: HTMLElement, hapticType: string)`.

Then implementation is `forwardHaptic = (el, type) => fireEvent(el, 'haptic_event', { type });`

Then inside external auth you can decide to add an event listener to body to translate it to external auth. That way you can also revert all the types ;)

However, this is all granted that this is a good idea, which I don't know yet if it is. If we implement it, we need to add documentation on when to implement it and where. Right now it's inside UI components, when an action is received, just before we fire off a service. That is probably an ok place. But what about failures? Is it up to the service handler to do this?

Reply:
Focusing just on that last sentence until we get past it...

I think haptics are a "best case" situation. If everything works and the phone shakes, great. If not, it's not a major loss. I don't think we need to inform the device of failures in getting it haptic info. It's a simple enough system that I can't even imagine failure scenarios. Also, what did you mean by service handler?

Reply:
The service handler is the function that handles service calls, that is where you currently have added an error haptic signal.

Reply:
Ah yeah sure. I put it there only because that triggers the "Failure" notification in the bottom right of the UI, right?

Reply:
The toast. I wanted the haptic to be in sync with that notification so users notice it.

Reply:
makes sense

Reply:
So this is I guess an iOS only thing right ? Android users don't vibrate all the time.

Reply:
Android the OS doesn't have a taptic engine like iPhone but some devices do. I can add rudimentary support with the vibration motor to the official app (someday).

Reply:
@balloob should I understand that you are okay with this PR moving forward? If so I'll go ahead and do the requested changes.

Reply:
Yes